crates/primitives/src/log.rs

```diff
@@ -116,21 +116,49 @@ impl Log {
     }
 }
 
+impl<T> Log<T>
+where
+    for<'a> &'a T: Into<LogData>,
+{
+    /// Creates a new log.
+    #[inline]
+    pub const fn new_from_event_unchecked(address: Address, data: T) -> Self {
+        Self { address, data }
+    }
+
+    /// Creates a new log from an deserialized event.
+    pub fn new_from_event(address: Address, data: T) -> Option<Self> {
+        let this = Self::new_from_event_unchecked(address, data);
+        (&this.data).into().is_valid().then_some(this)
+    }
+
+    /// Reserialize the data.
+    #[inline]
+    pub fn reserialze(&self) -> Log<LogData> {
+        Log { address: self.address, data: (&self.data).into() }
+    }
+}
+
 #[cfg(feature = "rlp")]
-impl alloy_rlp::Encodable for Log {
+impl<T> alloy_rlp::Encodable for Log<T>
+where
+    for<'a> &'a T: Into<LogData>,
+{
```

Comment:
ah I see, can just be converted to Log\<LogData> when serializing

Reply:
right. and because we are _usually_ 

This is significantly less ergonomic with `DynSolEvent` tho, as a decoded sol event can't be assumed to be validly encoded the way a decoded `sol!` struct with `SolEvent` can be. That 
assumption allows the infallible reserialize

cc @DaniPopes do you think that we should attach the `DynSolEvent` as a prop on the `DecodedEvent` in dyn abi so that we can assume valid construction and encoding? This would let us `impl From<&DecodedEvent> for LogData`